### DISCUSSION TITLE
Import-based API for `astro:assets` (framework components)

### DISCUSSION THREAD

**Discussion by isaac-mcfadyen** [withastro/roadmap#666] [Proposal]
# Summary

`astro:assets` (experimental, Astro version `2.10.12`) works very well today inside `.astro` files. However, the experience is a bit lacking when using framework components (such as Svelte, Vue, Lit, etc). This is a proposal for an import-based API that allows use inside those framework tools.

# Background & Motivation

- `astro:assets` allows the use of props passed to the `<Image>` component to control the image parameters, such as `height`, `width`, and `format`.
- However, because `<Image>` is an Astro component, it can't be used in UI frameworks (Svelte, Preact, Lit, etc).
- Inside UI framework components, you can only use an import (`import Image from "../image.png"`) to import an image, which doesn't allow changing height/width/format or transforming the image at all.
- For a real-world example, I have a logo that is very large (1200px wide) used in many places on my site, including inside Svelte components. Right now I use `vite-imagetools` to import the same logo many times with different widths/heights and formats (for example, the header only uses a 120px wide logo while the body uses 500px).
- I can't use the Astro `Image` component in these Svelte components, and I can't use `vite-imagetools` alongside `astro:assets` (see below section).

# Goals
- Allow a way to specify the imported image's parameters such as height, width, and format inside UI framework components.
- Similar tools such as `vite-imagetools` make this framework-agnostic by using query parameters on the import itself to specify the height, width, format, etc.
	- For example: `import Logo from "../Logo.png?h=120&w=120&format=avif";`
	- `vite-imagetools` is not compatible (it breaks and returns undefined for imports like this) when experimental `astro:assets` is enabled, so using both `astro:assets` and `vite-imagetools` at the same time is not an option.
- If a similar API (import-based transformations) were added to `astro:assets`, it could work alongside the existing `Image` component - users could use the `Image` component normally, and if using framework components, could use import-based annotations instead. 

# Example

For example, a Svelte component using this proposed API could look like:
```svelte
<script>
    import Logo from "../Logo.png?h=120&format=avif";
</script>

<img src={Logo.src} />
```

Notice that the above component looks very similar to the existing API as well as the API from `vite-imagetools` which is widely used across Vite-powered sites. On the imported image, there is `.src` and image is included in the final build. The only difference is that the imported image is automatically converted to `avif` and shrunk to `120px`, *exactly* the same as the Astro `Image` component would work but inside UI framework components.

**Comment by Princesseuh** [comment]
Yes! I wanted this something like day one of making `astro:assets` and would have made this exact proposal myself eventually, ha!

My only concerns are the following:
- What happens if you're using a remote image service?
- Can we do types properly? `vite-imagetools` suffer from this a lot in TypeScript, where every error becomes a type error because TypeScript does not handle having multiple wildstars in a pattern.

**Reply by isaac-mcfadyen** [reply]
My thoughts:
1. Unsure - does TS even support importing from a URL? Like could you `import Logo from "https://example.com/Logo.png?w=120"` and if so, what happens if the resource actually needs query parameters to fetch it? Would the transformation parameters be merged? Not sent? Something else?

2. Yeaaah this is a big issue with `vite-imagetools` I've seen too. My workaround for now is to:
```ts
// env.d.ts
declare module "*&imagetools" {
	var content: string;
	export default content;
}

// component.svelte
import Logo from "../Logo.png?w=120&imagetools"
```
This works because `vite-imagetools` ignores any unknown query params, so the `&imagetools` part is ignored by Vite but picked up by TypeScript. Note that TS doesn't support more than one wildcard, so you can't do `*.png?*` for example (see https://github.com/microsoft/TypeScript/issues/38638).

This also feels a bit hacky for using in Astro - what if a user wants to specify more than one format to get back a srcset or something? For example, `import Logo from "../Logo.png?w=120;240"` returns a `string[]` in `vite-imagetools` and I can't really think of a way to differentiate them in TS.

Final thought: could import assertions work eventually? They're not quite in TypeScript yet (see https://github.com/tc39/ecma262/pull/3057 for the initial proposal and https://github.com/microsoft/TypeScript/issues/53656 for the TypeScript support) but it might work nicely once they do get added/stabilized. Looks like it's targeted for TS 5.3 (https://github.com/microsoft/TypeScript/issues/55486)

**Comment by isaac-mcfadyen** [comment]
I saw Astro's new `<Picture>` components shipped today - looks amazing!

Wondering if there's any feedback/thoughts on how we could move this proposal forward? Not being able to use images in framework components (either via Astro's built-in image utilities or via `vite-imagetools`, which breaks in the newer versions of Astro) is somewhat of a blocker.

**Reply by Princesseuh** [reply]
`vite-imagetools` should work just fine on latest versions of Astro, Astro skips processing on imports with query params

**Reply by isaac-mcfadyen** [reply]
Apologies - you are correct! I've been using an old version of Astro, apparently 😅
